File: frogment/src/main/java/ch/grze/frogment/FrogmentFoundation.java
package ch.grze.frogment;

import android.app.Application;

import java.util.ArrayList;
import java.util.List;

import ch.grze.frogment.core.Config;
import ch.grze.frogment.core.Core;
import ch.grze.frogment.core.extension.AbstractExtension;

public final class FrogmentFoundation {
    private final Core core;

    public FrogmentFoundation(Application application) {
        this(new Builder(application));
    }

    private FrogmentFoundation(Builder builder) {
        core = new Core(builder.application, builder.config, builder.extensions);
    }

    public static class Builder {
        private final List<AbstractExtension> extensions = new ArrayList<>();
        private final Application application;
        private Config config = Config.getDefault();

        public Builder(Application application) {
            this.application = application;
        }

        public Builder config(Config config) {
            this.config = config;
            return this;
        }

        public Builder extension(AbstractExtension extension) {
            extensions.add(extension);
            return this;
        }

        public FrogmentFoundation build() {
            return new FrogmentFoundation(this);
        }
    }
}

File: frogment/src/main/java/ch/grze/frogment/FrogmentFoundation.kt
package ch.grze.frogment

import android.app.Application
import ch.grze.frogment.core.Config
import ch.grze.frogment.core.Core
import ch.grze.frogment.core.extension.AbstractExtension
import java.util.*

class FrogmentFoundation private constructor(builder: Builder) {
    private val core: Core

    constructor(application: Application) : this(Builder(application))

    init {
        core = Core(builder.application, builder.config, builder.extensions)
    }

    class Builder(val application: Application) {
        internal val extensions = ArrayList<AbstractExtension>()
        internal var config = Config.getDefault()

        fun extension(extension: AbstractExtension): Builder {
            extensions.add(extension)
            return this
        }

        fun config(config: Config): Builder {
            this.config = config
            return this
        }

        fun build(): FrogmentFoundation = FrogmentFoundation(this)
    }
}
